### PR TITLE
autocase OCP-48610

### PR DESCRIPTION
**Pull request by WenXinWei**

### CODE REVIEW

File: ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+unset SSL_CERT_FILE
+unset GOVC_TLS_CA_CERTS
+
+CLUSTER_NAME="${NAMESPACE}-${UNIQUE_HASH}"
+rsp_name="${CLUSTER_NAME}-rsp"
+
+if [[ -n "$(govc ls /${vsphere_datacenter}/host/${vsphere_cluster}/Resources/${rsp_name})" ]];then
+    echo "customized resource pool found, will delete."
+else
+    echo "cusomized resource pool not found.skip deprovision"
```
Comment (jianlinliu) [inline ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh:22]: Per my understanding, if users specified the customized resource pool, after openshift-install destroy clusters, the pool should exist, or else, it should fail the step to indicate there is some problem, right?

File: ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+unset SSL_CERT_FILE
+unset GOVC_TLS_CA_CERTS
+check_result=0
+
+machineset_name=$(oc get machineset -n openshift-machine-api --no-headers | head -n1 | awk '{print $1}')
+
+rsp_config=$(yq-go r ${INSTALL_CONFIG} 'platform.vsphere.failureDomains[*].topology.resourcePool')
+
+if [[ -n "$(oc get machineset -n openshift-machine-api ${machineset_name} -o json | jq -r .spec.template.spec.providerSpec.value.workspace | grep ${rsp_config})" ]];then
```
[the remark this answers is not in the document]
Comment (WenXinWei) [reply ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh:18]: seems we can not list machine/vm in resource pool. only can check on vSphere client @jinyunma  any suggestion?

File: ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+rsp_name="${CLUSTER_NAME}-rsp"
+
+if [[ -n "$(govc ls /${vsphere_datacenter}/host/${vsphere_cluster}/Resources/${rsp_name})" ]];then
+    echo "customized resource pool found, will delete."
+else
+    echo "cusomized resource pool not found.skip please check"
+    exit 1
+fi
+
+govc pool.destroy /${vsphere_datacenter}/host/${vsphere_cluster}/Resources/${rsp_name}
```
Comment (jianlinliu) [inline ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh:26]: Shall we move this line to the above `if` block ?

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.17-vsphere-ipi-customized-resourcepool-f28

File: ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh (hunk 3 above)
Comment (WenXinWei) [reply ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh:26]: ok

File: ci-operator/step-registry/cucushift/installer/rehearse/vsphere/ipi/customized-resourcepool/provision/cucushift-installer-rehearse-vsphere-ipi-customized-resourcepool-provision-chain.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+chain:
+  as: cucushift-installer-rehearse-vsphere-ipi-customized-resourcepool-provision
+  steps:
+  - chain: ipi-conf-vsphere
+  - ref: ipi-conf-vsphere-customized-resourcepool
+  - ref: ipi-conf-vsphere-folder
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/vsphere/ipi/customized-resourcepool/provision/cucushift-installer-rehearse-vsphere-ipi-customized-resourcepool-provision-chain.yaml:6]: Do you want to cover customized folder in this install workflow?

Comment (WenXinWei) [reply ci-operator/step-registry/cucushift/installer/rehearse/vsphere/ipi/customized-resourcepool/provision/cucushift-installer-rehearse-vsphere-ipi-customized-resourcepool-provision-chain.yaml:6]: no. forgot to delete the step. thanks for reminding.

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.17-vsphere-ipi-customized-resourcepool-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.17-vsphere-ipi-customized-resourcepool-f28

File: ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh:18]: From webconsole, can we see the created machines are in the resource pool, right?

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.17-vsphere-ipi-customized-resourcepool-f28

File: ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (jinyunma) [reply ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh:18]: We can also check vm list in resource pool, see [function](https://gitlab.cee.redhat.com/aosqe/v4-installer-autotest/-/blob/master/cases/TestCases/OCP-48610/vsphere_ipi_install_configurable_resourcepool.py?ref_type=heads#L37-45) in automated case OCP-48610 in our v4-installer-autotest repo.

File: ci-operator/step-registry/ipi/conf/vsphere/customized-resourcepool/ipi-conf-vsphere-customized-resourcepool-commands.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+
+CONFIG="${SHARED_DIR}/install-config.yaml"
+PATCH="${SHARED_DIR}/resourcepool.yaml.patch"
+
+CLUSTER_NAME="${NAMESPACE}-${UNIQUE_HASH}"
+rsp_name="${CLUSTER_NAME}-rsp"
+govc pool.create ${vsphere_cluster}/Resources/${rsp_name}
+
+cat > "${PATCH}" << EOF
```
Comment (jinyunma) [inline ci-operator/step-registry/ipi/conf/vsphere/customized-resourcepool/ipi-conf-vsphere-customized-resourcepool-commands.sh:24]: Actually, 4.9+ CI jobs are always running with specifying custom resource pool, see [code here](https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/conf/vsphere/ipi-conf-vsphere-commands.sh#L125).

maybe we can only add post-check here, and if resourcePool is defined in install-config, then check it. In deprovision step, just check that cluster machines are removed from resource pool after cluster is destroyed.

File: ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (WenXinWei) [reply ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh:18]: got it.

File: ci-operator/step-registry/ipi/conf/vsphere/customized-resourcepool/ipi-conf-vsphere-customized-resourcepool-commands.sh (hunk 5 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/vsphere/customized-resourcepool/ipi-conf-vsphere-customized-resourcepool-commands.sh:24]: Good point, agree.

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-template-f28

File: ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+set -o pipefail
+
+# shellcheck source=/dev/null
+source "${SHARED_DIR}/govc.sh"
+unset SSL_CERT_FILE
+unset GOVC_TLS_CA_CERTS
+INSTALL_CONFIG="${SHARED_DIR}/install-config.yaml"
+
+rsp_config=$(yq-go r ${INSTALL_CONFIG} 'platform.vsphere.failureDomains[*].topology.resourcePool')
+if [[ -z "$(govc pool.info -json ${rsp_config} | jq -r '.ResourcePools[].Vm[] | join(":")' | xargs govc ls -L | awk -F'/' '{print $NF}')" ]];then
```
Comment (jinyunma) [inline ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh:13]: Resource pool specified in install-config is shared to all vSphere clusters, when you run this command, all machines created by CI jobs can be listed.
```
# govc pool.info -json /DEVQEdatacenter/host/DEVQEcluster/Resources/ipi-ci-clusters | jq -r '.ResourcePools[].Vm[] | join(":")' | xargs govc ls -L | awk -F'/' '{print $NF}'
ci-op-k83sn59i-182b3-h5sgz-worker-0-jwmqg
ci-op-k83sn59i-182b3-h5sgz-worker-0-56qdq
ci-op-k83sn59i-182b3-h5sgz-worker-0-9rp5q
ci-op-k83sn59i-182b3-h5sgz-master-1
ci-op-k83sn59i-182b3-h5sgz-master-2
ci-op-k83sn59i-182b3-h5sgz-master-0
ci-op-x08l6vxc-bab81-qprn2-worker-0-hcr8r
ci-op-x08l6vxc-bab81-qprn2-master-0
ci-op-x08l6vxc-bab81-qprn2-worker-0-tqw5t
ci-op-x08l6vxc-bab81-qprn2-master-2
ci-op-x08l6vxc-bab81-qprn2-master-1
ci-op-x08l6vxc-bab81-bastion
ci-op-x08l6vxc-bab81-qprn2-worker-0-bpsxb
```
At least you need to append filtering with `infraId` or `clusterName` + excluding bastion host to check if related machines are destroyed or not, else this step might be always failed.

File: ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+export KUBECONFIG=${SHARED_DIR}/kubeconfig
+# shellcheck source=/dev/null
+source "${SHARED_DIR}/govc.sh"
+INSTALL_CONFIG="${SHARED_DIR}/install-config.yaml"
+unset SSL_CERT_FILE
+unset GOVC_TLS_CA_CERTS
+check_result=0
+
+machineset_name=$(oc get machineset -n openshift-machine-api --no-headers | head -n1 | awk '{print $1}')
+rsp_config=$(yq-go r ${INSTALL_CONFIG} 'platform.vsphere.failureDomains[*].topology.resourcePool')
```
Comment (jinyunma) [inline ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh:15]: It's better to check if `rsp_config` is empty, and skip the check and exit 0 when it is empty.

File: ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh (hunk 6 above)
Comment (WenXinWei) [reply ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh:13]: oh, you're right. will update.

File: ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh (hunk 7 above)
Comment (WenXinWei) [reply ci-operator/step-registry/cucushift/installer/check/vsphere/customized-resourcepool/cucushift-installer-check-vsphere-customized-resourcepool-commands.sh:15]: ok.

**Comment by WenXinWei** [comment]
pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installation-nightly-4.17-vsphere-ipi-customized-resourcepool-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

File: ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-ref.metadata.json [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{
+	"path": "vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-ref.yaml",
```
Comment (jianlinliu) [inline ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-ref.metadata.json:2]: Actually this step is not deprovision resource pool, but run checking, so need to rename the step name.

File: ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+INSTALL_CONFIG="${SHARED_DIR}/install-config.yaml"
+INFRA_ID=$(jq -r .infraID ${SHARED_DIR}/metadata.json)
+rsp_config=$(yq-go r ${INSTALL_CONFIG} 'platform.vsphere.failureDomains[*].topology.resourcePool')
+
+if [[ -z ${rsp_config} ]];then
+    echo "resource pool not defined in install-config. skip this step"
+    exit 0
+fi
+
+if govc pool.info -json ${rsp_config} | jq -r '.ResourcePools[].Vm[] | join(":")' | xargs govc ls -L | awk -F'/' '{print $NF}' | grep -q ${INFRA_ID};then
```
Comment (jianlinliu) [inline ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh:18]: I think here is two check points:
1. After clusters are destroyed, the resource pool still exists.
2. There is no vm in the resource pool.

It is better to run two separate if block to run the checks.

Comment (WenXinWei) [reply ci-operator/step-registry/vsphere/deprovision/customized-resourcepool/vsphere-deprovision-customized-resourcepool-commands.sh:18]: ok

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

**Comment by WenXinWei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-ipi-proxy-fips-f28

File: ci-operator/step-registry/vsphere/deprovision/customized-resourcepool-check/vsphere-deprovision-customized-resourcepool-check-commands.sh [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+if [[ -z ${rsp_config} ]];then
+    echo "resource pool not defined in install-config. skip this step"
+    exit 0
+fi
+
+if [[ -n "$(govc ls ${rsp_config})" ]];then
+    echo "customized resource pool ${rsp_config} was not deleted after cluster destroyed. that is expected"
+else
+    echo "customized resource pool ${rsp_config} not exists after cluster destroyed, please check"
```
Comment (jinyunma) [inline ci-operator/step-registry/vsphere/deprovision/customized-resourcepool-check/vsphere-deprovision-customized-resourcepool-check-commands.sh:21]: should `exit 1` if customized resource pool does not exist?

**Comment by jianlinliu** [comment]
lgtm

leave this to @jinyunma for final confirm.

**Pull request merged**